### ISSUE TITLE
Make audio a wrapper over audio-buffer and audio-buffer-utils

### ISSUE THREAD

**Issue by dfcreative** [audiojs/audio#22]
Hi @jamen!
Do I understand right that audio is opinionated container for audio-data with some handy API methods? Is there any sense to pivot it from audio-buffer so to provide a bunch of common methods for audio manipulations, e. g. all from [audio-buffer-utils](https://github.com/jaz303/audio-buffer-utils)?
That would be handy to require audio one time and have all the manip work done on some audio-buffer, I sould use that in [wavearea](https://github.com/audio-lab/wavearea) as well.
Using audio-buffer and utils right now is a bit wearisome.

**Comment by jamen** [comment]
That does sound good!  Would you like to PR this or should I?

v2 of `audio` is not released yet, it is currently in prerelease.  I think it is worth using that version for when it is decided we do release.

**Comment by jamen** [comment]
I'll add you to `audio` on npm as well.

**Comment by dfcreative** [comment]
Btw it appears that using node’s `Buffer['writeInt']` etc methods are 3-5 times slower than having a `TypedArray` instance created. I guess replacing `source` object with `audio-buffer` instance is a right move?

**Comment by jamen** [comment]
I did want to move away from `Buffer` entirely at one point.  So I guess it is haha.  Nice work :+1:

**Comment by dfcreative** [comment]
Started working on refactoring - I am afrad I have to remove almost all code, because it is covered by audio-buffer or audio-buffer-utils...
Also I wonder - how much `read` and `write` methods are really useful for a single-sample picking? I strongly feel like having a function call for getting a sample value is bad for performance. 
Also I am a bit confused with `write(x, offset, channel)` - it makes channels uneven, right?
I would suggest extending them to `write(values, offset?, channel?)` and `read(offset, size)`

**Comment by dfcreative** [comment]
Btw practically there is one nuance. It appears that operations on big data, like 1M+ samples are way slower than on small array. That is very bad for RT nature of audio processing (see [gl-waveform](https://audio-lab.github.io/gl-waveform/) after 1min of playing).
Therefore I would suggest a webworker mode, which would make methods asynchronous, but allow for instant data storing, which would not block main UI/audio thread.
I was planning to do that, as some components share waveform buffer, but I didn’t have that object, therefore there is a change for _audio_ be that awesome class.
Thoughts?

**Comment by jamen** [comment]
I'm fine with most of the code being replaced if it is for the better.  There hasn't been much testing or components made around this module so I expect some things aren't well designed.

Also, with writing single values my intent was for them to be used in loops.  In the previous version using `write(values, ...)` with an array was just doing a loop behind the scenes.  In my experience though functions calls are expensive when it comes down to performance, so you are probably right.

With `write(x, offset, channel)`, the `offset` is the block (which contains the different channels' data) and then `channel` is `1`, `2`, `3`, etc.  the time it is uneven is if you exceed the `audio.channels` number, and I probably should have had an if statement for preventing overflows, whoops.

Your ideas sound awesome, but that is new stuff for me so I wouldn't know how to approach that.  Audio is kind of my aspiring interest, so this is all a cool new opportunity for me. :)

**Comment by jamen** [comment]
I've seen some cool new stuff with [SIMD](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SIMD) by the way.  Not sure if you've seen/used it already, but it looks like something we might be able to use.  (Although it is an ES7 thing)

**Comment by dfcreative** [comment]
Looks sweet, definitely it is worth trying. Interesting how good is that for audio.
Nice [SIMD performance research](https://blog.mozilla.org/javascript/2015/03/10/state-of-simd-js-performance-in-firefox/).

A couple of things though.
First, RT audio deals with frequent read/write operations, as it puts chunks to memory and disposes them, but that is a weak point of SIMD, basically it is the overhead it adds over TypedArrays.
Second, audio data is causal, i. e. value of next sample depends on the value of previous one, that is the problem why audio data is not that easily processed in parallel, including the GPU as well. 
Third, SIMD is grouped by blocks, which might be an issue in case of buffer editing, but not that big.
So SIMD is good for image/large data processing/mining, but it is interesting what are good applications for audio. Maybe some convolution?

**Comment by jamen** [comment]
Perhaps the grouping could be useful in iterative opertions.  Like instead of 1 at a time, use 4 or 8 at a time?  Or am I thinking of that wrong?

Also, one concern I just had, are components supposed to use `AudioBuffer` or `Audio` now?

**Comment by dfcreative** [comment]
Yeah, but not all the iterative cases. Eg 'x[i] = x[i-1] * n' - it can't be done in parallel.
And Audio is supposed to be a high-level wrapper, for user. Sure it can be used in components, if it is reasonable

**Comment by dfcreative** [comment]
I think of audio.js as jQuery but for audio. Possible ideas of applications:

* [ ] _write_ data (array, buffer or audio-buffer) to the end or to some offset, actual for live recording
* [ ] _read_ slice of data, as array buffer, e. g. selected area
* [ ] _load_ file by URL, through xhr - very handy instead manual procedure. Therefore, `load` event.
* [ ] _stream from_, to save to file or send to speaker, in pcm format of choice or as audio-buffers sequence.
* [ ] _stream to_, to record stream of data into itself, from mic for example.
* [ ] _download_ - generate download link for a file of specific format, like [opentype.js](https://github.com/nodebox/opentype.js) does for font. Easier, than trying to think up how to manually save audio as file.
* [ ] _normalize_, _reverse_, _trim_ etc - micro audio manipulation methods from audio-buffer-utils.
* [ ] _map_ to apply amplitude modulation etc.
* [ ] _map channels_  to change channels layout/split/merge.
* [ ] _map-reduce_ to generate snapshot of grouped data by bars for waveform viewers.
* [ ] _frequencies_ for getting frequencies domain data.
* [ ] _fadeIn_, _fadeOut_ to add fades at the end/start or at some selected slice.
* [ ] _stretch_ to change pitch.
* [ ] _tone_ detection/setup, to align pitch with other audio, in case of merge.
* [ ] _tempo_ detection/setup, to align tempo in combination with other audio.
* [ ] _mix_ to mix properly, not just average with other audio, i. e. with overlapping.
* [ ] _effects_ hooks, for plugins, like noise reduction, eq, filter, etc - basically, any stream processing.
* [ ] _meta_ fields, which possibly will be serialized to target format.
* [ ] _worker_ mode to bring data-operations and store data in the background.

**Comment by dfcreative** [comment]
Good to look at [howl.js](https://github.com/goldfire/howler.js/#quick-start)

**Comment by jamen** [comment]
Should we make a lot of that functionality as components if not already, then wrap it together here?

**Comment by jamen** [comment]
Here is something I found that could help with the webworker mode: https://github.com/maxogden/workerstream

**Comment by dfcreative** [comment]
Yeah, with separate packages, as far as it is reasonable.

As for webworker it is very nice idea. Here I guess we can get along with webworkify only, but for streams we should definitely try. Maybe try to place subprocessing pipeline into webworker stream.

**Comment by dfcreative** [comment]
For example  we have `audio.load(src, cb?)` method. Should we return promise or self instance from it?
```js
//Promise:
audio.load('./sound.mp3').then(
result => {
},
error => {
});

//Chain 
audio.load('./sound.mp3').once('load', (err, result) => {
});
```
I like the second, it is classical way and does not break chain of calls. Also promises are hellish in dev/support tbh.

**Comment by jamen** [comment]
Chaining sounds good to me as well. :)

**Comment by jamen** [comment]
@dfcreative Do you think it is worth putting a notice in the README until we create something compatible with all the other audio modules?

**Comment by dfcreative** [comment]
Mb. I started refactoring in [class branch](https://github.com/audiojs/audio/tree/class), implemented [audio-play](https://github.com/audiojs/audio-play) for playback API and [audio-decode](https://github.com/audiojs/audio-decode) for node. Waiting for resolution of _audio-play_ name in npm and @danigb to make [audio-loader](https://github.com/danigb/) work in node, then we have almost all needed to implement audio@3.0.0.

**Comment by dfcreative** [comment]
API reference https://github.com/jiaaro/pydub/blob/master/API.markdown

**Comment by dfcreative** [comment]
API gotchas.
1. How to invert phase of only right channel?
2. How to read raw offsets rather then seconds?
3. How to fade decibels?

In advanced programming languages it is done with atoms or named arguments:
```js
audio.invert(from=0s, to=1s, channel=1)
audio.data(from=100, to=8000, channel=left)
audio.fade(from=1100ms, duration=500ms, gain=-20db)
```
What are possible workarounds for js?

**Comment by dfcreative** [comment]
So here is some js style ethics: https://codeutopia.net/blog/2016/11/24/best-practices-for-javascript-function-parameters/ for named arguments.
Also we can agree upon units, say, seconds and decibels.

```js
audio.invert({channel: 1})
audio.data({from: 100, to: 8000, raw: true, channel: 0})
audio.fade(1.1, .5, {gain: -20})
```
